Add experimental support for wrangler dev/publish file.wasm

kkoenig opened the pull request:
Allow wrangler to work with simple with WebAssembly programs that
  operate on stdin/stdout by generating a js shim that uses
  '@cloudflare/wasi'

  Example:
      cargo new hello_wasm
      cd ./hello_wasm
      cargo build --target wasm32-wasi --release
      wrangler dev target/wasm32-wasi/release/hello_wasm.wasm
      curl http://localhost:8787
        Hello, world!

kkoenig commented:
cc @Electroid @threepointone @petebacondarwin

tniessen commented on packages/wrangler/src/wasi.ts:
```diff
@@ -0,0 +1,70 @@
+export async function wrapIfWasmEntrypoint(
+  entrypoint: string
+): Promise<string> {
+  if (path.extname(entrypoint) != ".wasm") {
```
```suggestion
  if (path.extname(entrypoint) !== ".wasm") {
```

tniessen commented on packages/wrangler/src/wasi.ts:
```diff
@@ -0,0 +1,70 @@
+  const args = process.env[argsPrefix]?.split(',') ?? []
+
+  // Allow setting env variables by putting WRANGLER_WASM_ENV_xyz in wrangler
+  // envinronment, eg. WRANGLER_WASM_ENV_FOO=baz -> FOO=baz
```
```suggestion
  // environment, e.g., WRANGLER_WASM_ENV_FOO=baz -> FOO=baz
```

tniessen commented on packages/wrangler/src/wasi.ts:
```diff
@@ -0,0 +1,70 @@
+  const name = path.basename(wasmPath);
+
+  // Allow setting cli arguments by putting WRANGLER_WASM_ARGS=foo=bar,baz in wrangler
+  // envinronment, eg. WRANGLER_WASM_ARGS=foo=bar,baz -> hello.wasm foo=bar baz
```
```suggestion
  // environment, e.g., WRANGLER_WASM_ARGS=foo=bar,baz -> hello.wasm foo=bar baz
```

geelen commented:
This is coooooool! As... someone who hasn't even dipped a toe into WASI yet, what kind of projects will this be enough to support straight off the bat. I saw @byule demo SWC, did that take any extra code or was this enough as-is?

threepointone commented on packages/wrangler/src/wasi.ts:
```diff
@@ -0,0 +1,70 @@
+
+// TODO:
+// - wrangler --local is broken due to different bundling behavior between 'ModuleCollector' for
+//   Remote and miniflare cli for Local
```
working on this in https://github.com/cloudflare/wrangler2/pull/334

threepointone commented:
This is looking pretty good so far! I'll do a proper review soon. Some quick notes - 
- Could you also add an `example-wasi-app` under `packages` so we can manually test this whenever? 
- Do you suppose we should do this for `.wasi` extension types only? Or would that expect too much from the ecosystem? I don't know myself, hence asking. 
- We should think about tests for this, if only for publish for now.
- I'd like to avoid creating the tmp directory just for wrapping the entrypoint. We have a similar situation with the `static-asset-facade`, we should look to use a similar approach. I'll give it a go myself.

kkoenig commented:
> This is coooooool! As... someone who hasn't even dipped a toe into WASI yet, what kind of projects will this be enough to support straight off the bat. I saw @byule demo SWC, did that take any extra code or was this enough as-is?

I think it requires a little bit of rust plumbing, here is another demo courtesy of @zebp that works out of the box
```bash
git clone https://github.com/sharkdp/hexyl.git
cd ./hexyl
cargo build --release --target wasm32-wasi
wrangler2 dev ./target/wasm32-wasi/release/hexyl.wasm
echo "Hello world!" | curl http://localhost:8787 -X POST --data-binary @-
```

![image](https://user-images.githubusercontent.com/1980353/152216402-cd7a3c01-599a-4d7b-84b2-059ca2758ec1.png)

kkoenig commented:
> * Could you also add an `example-wasi-app` under `packages` so we can manually test this whenever?

Happy to do this, what were you thinking? I think the simplest thing would be to run `cargo new hello_wasm` and check in the source which would just be main.rs that prints hello world, but not sure if that is super useful or not

kkoenig commented:
> * Do you suppose we should do this for `.wasi` extension types only? Or would that expect too much from the ecosystem? I don't know myself, hence asking.

I dont think so, as far as I can tell `.wasi` is not really used anywhere.  In general I think we should try to stay consistent with `wasmtime run <file.wasm>` where it makes sense.  If folks get confused by trying to publish wasm modules that dont have the right exports we should be able to load the wasm module in wrangler prior to upload and check exports to give nice error messages

threepointone commented:
A rust project is good, but something that accepts a request and spits out "Hello World", I think. 

(We'll add an `example-wasm-app` separately that has explicit bindings (via a `workers-rs` sample app), but the `wasi` case is different ofc. )

> I dont think so, as far as I can tell .wasi is not really used anywhere. 

Noted noted. Thanks!

threepointone commented:
(or am I misunderstanding the difference between wasm and wasi?)

Electroid commented on packages/wrangler/package.json:
```diff
@@ -37,6 +37,7 @@
   ],
   "dependencies": {
     "esbuild": "0.14.14",
+    "@cloudflare/workers-wasi": "0.0.4",
```
Do we need to import this as a dependency? This is only used in the glue JS code that is generated, and it would be nicer if the developer could control the version of this dependency without wrangler pinning it.

Might be worth using the package manger that @petebacondarwin is introducing in #373?

Not a dependency expert, so also cc'ing @threepointone

kkoenig replied:
I'm not so sure we want to change this because for wasm there is no package.json, no wrangler.toml etc. that the consumer provides, where would they define which version they want to pin?

I'm open to exposing something like that but I think we should start with keeping the version controlled by wrangler so we can test end to end (with wrangler dev/publish etc.) when we bump the wrangler pinned version

Distributing it this way means theres no special thing that we have to worry about for wasi specifically, any case where wrangler is installed the module will be correctly installed and we can guarantee that the on-the-fly bundling will work.  Since the module is ~75k and likely to grow smaller over time (and has no prod deps) I don't think there is much downside

Electroid replied:
Good points.. been thinking so long in a JS world, glossed over not having the package.json et al.

Electroid commented on packages/wrangler/src/wasi.ts:
```diff
@@ -0,0 +1,70 @@
+
+  // Allow setting cli arguments by putting WRANGLER_WASM_ARGS=foo=bar,baz in wrangler
+  // environment, e.g., WRANGLER_WASM_ARGS=foo=bar,baz -> hello.wasm foo=bar baz
+  const argsPrefix = "WRANGLER_WASM_ARGS";
```
```suggestion
  const argsPrefix = "WRANGLER_WASI_ARGS";
```
Should we make this more specific to WASI?

kkoenig replied:
I don't have a strong preference (but I don't think WASM is ambiguous because in what other context would args/env variables be passed to wasm?)

I would consider this config an experimental placeholder, once we get folks playing around with this I'd like to meet and discuss the appropriate cli interface (ideally we get close to `wasmtime run`)

kkoenig replied:
I actually think we should avoid using the WASI terminology where possible

kkoenig commented:
> A rust project is good, but something that accepts a request and spits out "Hello World", I think.

Not quite sure I understand but we can chat more offline.  The "example" would look like this:
```
cargo new hello_wasm
cd ./hello_wasm
cargo build --target wasm32-wasi --release
wrangler dev target/wasm32-wasi/release/hello_wasm.wasm
```

In the rust source there would be no request handling or response handling, this would be the contents of main.rs
```rust
fn main() {
    println!("Hello, world!");
}
```

All the scaffolding etc. gets handled by the on-the-fly wrangler bundling that pulls in workers-wasi and generates index.js, so the paradigm is a bit different than existing examples

threepointone commented:
I haven't been able to get this to actually work locally. I end with an error that looks like this - 
![image](https://user-images.githubusercontent.com/18808/153567442-1ff00a67-4082-44f2-b315-f14a1af9cebe.png)
I made a hello_wasm package following the instructions at `packages/hello_wasm` (and added a package.json, and ran npm install at the root)

kkoenig commented:
> I haven't been able to get this to actually work locally. I end with an error that looks like this - ![image](https://user-images.githubusercontent.com/18808/153567442-1ff00a67-4082-44f2-b315-f14a1af9cebe.png) I made a hello_wasm package following the instructions at `packages/hello_wasm` (and added a package.json, and ran npm install at the root)

This looks like a problem on OSX vs linux, looking into it

Thanks for setting up the branch that will work great

kkoenig commented:
> I've also setup an experimental-wasm branch that includes this PR, rebased on main. Any commits to that branch get published as `wrangler@wasm`, but you'll see the error happens regardless.



> I've also setup an experimental-wasm branch that includes this PR, rebased on main. Any commits to that branch get published as `wrangler@wasm`, but you'll see the error happens regardless.

Added a new commit to this PR with the fix, let me know if I should create a PR to that branch instead or how you think we should manage it

threepointone commented:
It would be awesome if you could set up a PR for the experimental-wasm branch! I'll reset it to be HEAD right now

threepointone commented:
The branch now points to the latest commit, feel free to send a PR, thank you!

kkoenig commented:
> The branch now points to the latest commit, feel free to send a PR, thank you!

Ok should be all merged and I updated the base branch of this PR to the experimental-wasm branch

kkoenig commented:
Looks like 1 check is still failing with "Run marocchino/sticky-pull-request-comment@v2
Error: Resource not accessible by integration"

I think this will currently fail for all PRs from fork, so not related to this change

petebacondarwin commented:
> Looks like 1 check is still failing with "Run marocchino/sticky-pull-request-comment@v2 Error: Resource not accessible by integration"
> 
> I think this will currently fail for all PRs from fork, so not related to this change

Yes, you are right. I have a fix in the works but it is awkward.

threepointone commented:
I tested this, and it works! Let's land it and let it cook.
